Document convert_concl_no_check (#3225) (#10012)
coq/coq: doc/sphinx/proof-engine/tactics.rst

      make some proof steps explicit when refactoring a proof script
      to make it readable.

   .. tacv:: convert_concl_no_check @term

      For advanced usage. Similar to :n:`change @term`, but as an optimization,
      it skips checking that :n:`@term` is convertible to the goal.

      Recall the Coq kernel typechecks proofs again when they are concluded to

>       ensure safety. Hence, using :n:`change` checks convertibility twice

Reviewer: ```suggestion
      ensure safety. Hence, using :tacn:`change` checks convertibility twice
```

>       overall, while :n:`convert_concl_no_check` can produce ill-typed terms,

Reviewer: You should use `:tacn:` here as well, but you need first to add a `:name:` option to the `.. tacv::` directive, otherwise it won't be indexed (because variants aren't by default).

      make some proof steps explicit when refactoring a proof script
      to make it readable.

   .. tacv:: convert_concl_no_check @term
      :name: convert_concl_no_check

      For advanced usage. Similar to :n:`change @term`, but as an optimization,
      it skips checking that :n:`@term` is convertible to the goal.

      Recall the Coq kernel typechecks proofs again when they are concluded to
      ensure safety. Hence, using :tacn:`change` checks convertibility twice
      overall, while :tacn:`convert_concl_no_check` can produce ill-typed terms,
      but checks convertibility only once.
      Hence, :tacn:`convert_concl_no_check` can be useful to speed up certain proof
      scripts, especially if one knows by construction that the argument is
      indeed convertible to the goal.

      In the following example, :tacn:`convert_concl_no_check` replaces :g:`False` by
      :g:`True`, but :g:`Qed` then rejects the proof, ensuring consistency.

      .. example::

         .. coqtop:: all abort

            Goal False.
              convert_concl_no_check True.
              exact I.
              Fail Qed.

> .. tacn:: exact_no_check @term

Reviewer: The location of this documentation doesn't seem right. If you want to have all the `no_check` tactic be documented close together, then you could create a new section for these tactics. But here, you chose to document them in the section on reduction tactics, which only makes sense for `convert_concl_no_check`.

   :name: Mangle Names Prefix

   Specifies the prefix to use when generating names.

Performance-oriented tactic variants
------------------------------------

.. tacn:: change_no_check @term
   :name: change_no_check

   For advanced usage. Similar to :n:`change @term`, but as an optimization,

>    it skips checking that :n:`@term` is convertible to the goal.

Reviewer 2: Minor grammar nit (not strictly necessary, but I think it reads better):
```suggestion
   it skips checking that :n:`@term` is convertible with the goal.
```

Author: Your call, but I wouldn't. The manual prefers "A and B are convertible", followed by "convertible to", then "convertible with".

>    Recall the Coq kernel typechecks proofs again when they are concluded to

Reviewer 2: Minor grammar nit:
```suggestion
   Recall that the Coq kernel typechecks proofs again when they are concluded to
```

   ensure safety. Hence, using :tacn:`change` checks convertibility twice

>    overall, while :tacn:`convert_concl_no_check` can produce ill-typed terms,

Reviewer 2: ```suggestion
   overall, while :tacn:`change_no_check` can produce ill-typed terms,
```

   but checks convertibility only once.

>    Hence, :tacn:`convert_concl_no_check` can be useful to speed up certain proof

Reviewer 2: ```suggestion
   Hence, :tacn:`change_no_check` can be useful to speed up certain proof
```

   scripts, especially if one knows by construction that the argument is
   indeed convertible to the goal.

>    In the following example, :tacn:`convert_concl_no_check` replaces :g:`False` by

Reviewer 2: ```suggestion
   In the following example, :tacn:`change_no_check` replaces :g:`False` by
```

   :g:`True`, but :g:`Qed` then rejects the proof, ensuring consistency.

   .. example::

      .. coqtop:: all abort

         Goal False.
           change_no_check True.
           exact I.
         Fail Qed.

   .. tacv:: convert_concl_no_check @term
      :name: convert_concl_no_check

      Deprecated old name for :tacn:`change_no_check`. Does not support any of its
      variants.

.. tacn:: exact_no_check @term
   :name: exact_no_check

   For advanced usage. Similar to :n:`exact @term`, but as an optimization,
   it skips checking that :n:`@term` has the goal's type, relying on the kernel

>    check instead. See :tacn:`convert_concl_no_check` for more explanations.

Reviewer 2: ```suggestion
   check instead. See :tacn:`change_no_check` for more explanations.
```
